README
Osm2Pov - build 3D model from OpenStreetMap data!
Author Aleš Janda

----------
0. Content
----------

1. Introduction
2. Installation
3. Using
4. Some programator's stuff
5. Map scale
6. Program licence
7. How to do own map server
8. Contact


---------------
1. Introduction
---------------

The osm2pov is converter from OSM file (file in OpenStreetMap native format) to POV format (native format of POV-Ray renderer). So: you can download an OSM file from OpenStreetMap of your place of interest (or cut some area using Osmosis) and you can convert it to 3D model.

POV is format for POV-Ray but it's also language for full-fledged 3D models description. So it is possible to convert POV format to some 3D Modeller and walk in 3D city model. It's possible but I don't try it yet ;-)

Osm2Pov is written in C++.


---------------
2. Installation
---------------

Installation is from source code. It requires:
- g++ ver. 4.3 or greater (or some other C++ compiler)
- libexpat1-dev (for XML reading)


|| Compiling on Linux or other *nix systems: ||

$ cd projectdir/osm2pov
$ ./configure
$ make

If make failed, check if you really have libexpat1-dev, configure don't check it.
If compiler doesn't know -std=c++0x option, check if you have gcc at least 4.3. If you use other compiler, try remove this option from makefile.

Now you have a binary ./osm2pov in current directory.
If you want, you can install it:

# make install

Note for Mac OS X users: Unfortunately, standard version of gcc is 4.2 here. Try to proceed by this guideline: http://wiki.openstreetmap.org/wiki/Osm2pov_on_the_Mac


|| Compiling on Windows: ||

If you have Cygwin, you can compile osm2pov under it. Check that version of g++ is at least 4.3 (run g++ --version). Compilation is similar as in Linux.

If you don't want use Cygwin and you want native program, try this:
- open your favorite C++ IDE
- create new project
- add to it all *.cc files (*.cc file is the same as *.cpp; *.cc is only Linux convention)
- add libexpat1 as library to settings both to include dir and to link with program
- compile it

I don't have much experience with compile in Windows so good luck :-)


--------
3. Using
--------

osm2pov INPUT_FILE.osm OUTPUT_FILE.pov [X Y]

If converts OSM file INPUT_FILE.osm to POV-Ray file OUTPUT_FILE.pov.
X and Y are optionally and there are coords of zoom 12, where Y is divided by 2 (see ./osm2pov for details).

Using POV-Ray:

1) you must have file "osm2pov-styles.inc" and "textures" folder in current folder
2) download and install POV-Ray from Linux repositary or from http://www.povray.org/
3) convert POV to image using command like this: "povray +W1024 +H1024 +B100 +D +A +Imyplace.pov +Omyplace.png" - myplace.pov is input file, W/H is width/height of image (should be the same - square). If you have less memory, turn off thumbnail (type -D instead of +D). See povray documentation for more info
4) See the output image :-)


---------------------------
4. Some programator's stuff
---------------------------

The main file is osm2pov.cc. Function main() use three basic class
 - Primitives - as container for input data and reading input OSM file
 - Osm2PovConverter - object for converting input objects to 3D objects
 - PovWriter - for writing 3D objects to file in POV format

Styles for output images are in file "osm2pov-styles.inc". There are defined class used in output POV file. See POV-Ray documentation to explain it.
Tree images are in directory "textures".


------------
5. Map scale
------------

The output file has the same projection as Mapnik or Osmarenderer (Mercator) but in one exception - scale of Y axis is 2 times smaller than in Mapnik. It's because isometric map is viewed from 45 degrees, not 90 degrees as other maps. It causes that all the map is in weight 2 times smaller than in Mapnik.

This is why isn't possible to do simple "layer switch" between isometric and other map.


------------------
6. Program licence
------------------

This program is released under GPL v3. See http://www.gnu.org/licenses/gpl.html for licence details.


---------------------------
7. How to do own map server ...or make own image
---------------------------

Render one big map is rendering many small tiles. Osm2Pov converter is now dimensed to generate tiles in zoom 12. Due to asymetric scaling (see Map scale), zoom 12 is equal to two tiles of Mapnik zoom 12.

1) So, do the list of all [X,Y] coords in your place. X and Y are in zoom 12, Y divided by 2 (tiles count will 2 times smaller than in Mapnik, for example)

2) For each [X,Y] coord, get OSM file of this area. It's useful use Osmosis (http://wiki.openstreetmap.org/wiki/Osmosis) to do it:
osmosis --read-xml file=myarea.osm --bounding-box top=$MAXLAT left=$MINLON bottom=$MINLAT right=$MAXLON --write-xml file=myplace.osm

3) Every place (myplace.osm), convert it to POV file using osm2pov:
osm2pov myplace.osm myplace.pov $X $Y

4) Next, every POV file render using POV-Ray:
povray +W8192 +H8192 +B100 +FN -D +A +Imyplace.pov +Omyplace.png
If you want to see thumbnail, type +D instead -D. Size 8192x8192 is ideal for maximal zoom 17, if you want bigger/lesser zoom, multiple/divide sizes by 2 in each zoom level

5) Crop big PNG file into small tiles - you can using ImageMagick to this:
convert -quality 92 -crop 256x256 myplace.png tiles/tmp/xx.jpg
It also convert it to JPEG format. For resize, see "-resize" parameter. For montage, see "montage" command.

6) Move files into directory tree and you have map :-)

In folder "server-scripts" are some convert scripts for it. See it :-)


----------
8. Contact
----------

See http://osm.kyblsoft.cz/3dmapa/info to more details and contact.
Hope you enjoyed it ;-)



changelog.txt
Osm2Pov changelog:

For newest changes, see commits at https://github.com/guerda/osm2pov/commits/master or https://github.com/bitsteller/osm2pov

0.2.3 (2010-05-04)
 - partially merged with http://github.com/bitsteller/osm2pov - add support for trees, beaches, post-boxes, tram railways
 - improved style for railways
 - fixed output size when X and Y is at input (regression from 0.2.2)
 
0.2.2 (2010-05-03)
 - fixed scaling - output POV file is adaptable to size of input image so any input area size is now possible
 - styles: buildings with amenity=place_of_workship have black roof
 
0.2.1 (2010-04-28)
 - fixed rendering seams of riverways
 - water is now above forest floor

0.2 (2010-04-25)
 - support for tags in relation (not only tags in outer way) - tags in relation are preffered
 - support for multipolygons with more than one outer part

0.1.2 (2010-03-27)
 - fix some compilation issues

0.1.1 (2010-03-25)
 - fix running and output without coords X and Y

0.1 (2010-03-23)
 - initial version of program


osm2pov.cc
/**
 * Osm2Pov v0.1 - Copyleft Aleš Janda
 * This software is under GPL v3
 * See http://osm.kyblsoft.cz/3dmapa/info for more information
 */

#ifndef VERSION
 #define VERSION "undefined version"
#endif

#include <math.h>

#include "global.h"
#include "point_field.h"
#include "osm2pov_converter.h"
#include "output_polygon.h"
#include "pov_writer.h"
#include "primitives.h"


int main(int argc, const char **argv) {
	if (argc != 3 && argc != 5) {
		cout << "Osm2Pov " << VERSION;
		cout << "\tAuthor Aleš Janda | See http://osm.kyblsoft.cz/3dmapa/info for details" << endl << endl;
		cout << "Using:\t" << argv[0] << " input.osm output.pov [X Y]" << endl << endl;
		cout << "By default, area is computed from OSM file. If you can use it for render part of bigger map, set X and Y parameters. These are coords of tiles of zoom 12, where Y is divided by 2." << endl;
		cout << "Currently, zoom of output model (and image) is always the same." << endl;
		return 1;
	}

	Primitives primitives;
	bool fix_size_to_square = true;

	if (argc == 5) {
		primitives.setBoundsByXY(atol(argv[3]), atol(argv[4]));
		fix_size_to_square = false;
	}

	cout << "Loading input file" << endl;

	//setting attributes that are ignored when read OSM file. These attributes will not used (for memory saving)
	primitives.setIgnoredAttribute("addr:city", NULL);
	primitives.setIgnoredAttribute("addr:conscriptionnumber", NULL);
	primitives.setIgnoredAttribute("addr:country", NULL);
	primitives.setIgnoredAttribute("addr:housenumber", NULL);
	primitives.setIgnoredAttribute("addr:provisionalnumber", NULL);
	primitives.setIgnoredAttribute("addr:postcode", NULL);
	primitives.setIgnoredAttribute("addr:street", NULL);
	primitives.setIgnoredAttribute("addr:streetnumber", NULL);
	primitives.setIgnoredAttribute("admin_level", NULL);
	primitives.setIgnoredAttribute("alt_name", NULL);
	primitives.setIgnoredAttribute("bicycle", NULL);
	primitives.setIgnoredAttribute("complete", NULL);
	primitives.setIgnoredAttribute("created_by", NULL);
	primitives.setIgnoredAttribute("description", NULL);
	primitives.setIgnoredAttribute("dibavod:id", NULL);
	primitives.setIgnoredAttribute("is_in", NULL);
	primitives.setIgnoredAttribute("is_in:continent", NULL);
	primitives.setIgnoredAttribute("is_in:country_code", NULL);
	primitives.setIgnoredAttribute("FIXME", NULL);
	primitives.setIgnoredAttribute("foot", NULL);
	primitives.setIgnoredAttribute("full_name", NULL);
	primitives.setIgnoredAttribute("kct_blue", NULL);
	primitives.setIgnoredAttribute("kct_yellow", NULL);
	primitives.setIgnoredAttribute("maxspeed", NULL);
	primitives.setIgnoredAttribute("motorcycle", NULL);
	primitives.setIgnoredAttribute("motorcar", NULL);
	primitives.setIgnoredAttribute("name", NULL);
	primitives.setIgnoredAttribute("note", NULL);
	primitives.setIgnoredAttribute("name", NULL);
	primitives.setIgnoredAttribute("old_name", NULL);
	primitives.setIgnoredAttribute("operator", NULL);
	primitives.setIgnoredAttribute("place", NULL);
	primitives.setIgnoredAttribute("ref", NULL);
	primitives.setIgnoredAttribute("ref:bridge", NULL);
	primitives.setIgnoredAttribute("route", NULL);
	primitives.setIgnoredAttribute("source", NULL);
	primitives.setIgnoredAttribute("source:addr", NULL);
	primitives.setIgnoredAttribute("source:name", NULL);
	primitives.setIgnoredAttribute("source:loc", NULL);
	primitives.setIgnoredAttribute("source:position", NULL);
	primitives.setIgnoredAttribute("source:ref", NULL);
	primitives.setIgnoredAttribute("street", NULL);	//?
	primitives.setIgnoredAttribute("todo", NULL);
	primitives.setIgnoredAttribute("uir_adr:ADRESA_KOD", NULL);
	primitives.setIgnoredAttribute("voltage", NULL);
	primitives.setIgnoredAttribute("wifi", NULL);

	// set attributes that it remember but it will not be searched, only using as specify other tags
	// this all is for warnings what tags wasn't used
	primitives.setLightlyIgnoredAttribute("bridge", "yes");
	primitives.setLightlyIgnoredAttribute("building:height", NULL);
	primitives.setLightlyIgnoredAttribute("building:levels", NULL);
	primitives.setLightlyIgnoredAttribute("have_riverbank", "yes");
	primitives.setLightlyIgnoredAttribute("height", NULL);
	primitives.setLightlyIgnoredAttribute("layer", NULL);
	primitives.setLightlyIgnoredAttribute("type", "multipolygon");
	primitives.setLightlyIgnoredAttribute("wood", NULL);

	//loading from file
	if (!primitives.loadFromXml(argv[1])) return 1;

	cout << "Writing POV file" << endl;
	PovWriter pov_writer(argv[2], primitives.getViewRect(), fix_size_to_square);
	if (!pov_writer.isOpened()) return 1;

	Osm2PovConverter osm2pov_converter(&primitives, &pov_writer);

	//generating objects
	osm2pov_converter.drawWaysWithBorder("highway", "motorway", 10, 0.06, "highway", 10, "highway_secondary_border");
	osm2pov_converter.drawWaysWithBorder("highway", "motorway_link", 5.5, 0.06, "highway", 10, "highway_border");
	osm2pov_converter.drawWaysWithBorder("highway", "primary", 8, 0.06, "highway", 10, "highway_secondary_border");
	osm2pov_converter.drawWaysWithBorder("highway", "primary_link", 5.5, 0.06, "highway", 10, "highway_border");
	osm2pov_converter.drawWaysWithBorder("highway", "trunk", 7, 0.06, "highway", 10, "highway_secondary_border");
	osm2pov_converter.drawWaysWithBorder("highway", "trunk_link", 5, 0.06, "highway", 10, "highway_border");
	osm2pov_converter.drawWaysWithBorder("highway", "secondary", 7, 0.06, "highway", 10, "highway_secondary_border");
	osm2pov_converter.drawWaysWithBorder("highway", "tertiary", 6.5, 0.06, "highway", 10, "highway_border");
	osm2pov_converter.drawWaysWithBorder("highway", "unclassified", 6, 0.06, "highway", 10, "highway_border");
	osm2pov_converter.drawWaysWithBorder("highway", "road", 6, 0.06, "highway", 10, "highway_border");

	osm2pov_converter.drawWays("highway", "residential", 5, 0.06, "highway", true, true);
	osm2pov_converter.drawWays("highway", "living_street", 5, 0.06, "highway", true, true);
	osm2pov_converter.drawWays("highway", "service", 4, 0.06, "highway", true, true);
	osm2pov_converter.drawWays("highway", "pedestrian", 4, 0.05, "highway", true, true);

	osm2pov_converter.drawWays("highway", "track", 3, 0.045, "path", true, false);
	osm2pov_converter.drawWays("highway", "footway", 2, 0.03, "footway", true, false);
	osm2pov_converter.drawWays("highway", "steps", 2, 0.03, "footway", true, false);
	osm2pov_converter.drawWays("highway", "cycleway", 2.5, 0.03, "footway", true, false);
	osm2pov_converter.drawWays("highway", "path", 1.2, 0.015, "path", true, false);
	osm2pov_converter.drawAreas("amenity", "parking", 0.0009, "highway_area");

	osm2pov_converter.drawWays("railway", "rail", 5, 0.09, "railway", false, false);
	osm2pov_converter.drawWays("railway", "abandoned", 3, 0.09, "railway", false, false);
	osm2pov_converter.drawWays("railway", "disused", 3, 0.09, "railway", false, false);
	osm2pov_converter.drawWays("railway", "preserved", 3, 0.09, "railway", false, false);
	osm2pov_converter.drawWays("railway", "narrow_gauge", 3, 0.09, "railway", false, false);
	osm2pov_converter.drawWays("railway", "tram", 2.25, 0.12, "railway_tram", true, false);

	osm2pov_converter.drawAreas("leisure", "playground", 0.013, "playground");
	osm2pov_converter.drawAreas("leisure", "pitch", 0.013, "playground");

	osm2pov_converter.drawWays("aeroway", "runway", 40, 0.07, "highway", true, true);
	osm2pov_converter.drawWays("aeroway", "taxiway", 7, 0.07, "highway", true, true);

	osm2pov_converter.drawAreas("landuse", "cemetery", 0.012, "cemetery");
	osm2pov_converter.drawAreas("natural", "beach", 0.012, "beach");

	osm2pov_converter.drawBuildings("building", NULL, 4.5, { "building" }, { "building_living_roof1", "building_living_roof2", "building_living_roof3", "building_living_roof4" }, { "building_nonliving_roof1", "building_nonliving_roof2" }, { "building_religious_roof" });

	osm2pov_converter.drawSpecialBuildings("leisure", "stadium", 12, "man_made_tower", "man_made_tower");
	osm2pov_converter.drawSpecialBuildings("building:part", NULL, 3, "building", "building");
	osm2pov_converter.drawTowers("artwork_type", "obelisk", 4, 25, "man_made_tower"); //FIXME: for testing only
	osm2pov_converter.drawTowers("man_made", "tower", 4, 25, "man_made_tower");
	osm2pov_converter.drawTowers("amenity", "tower", 4, 25, "man_made_tower");
	osm2pov_converter.drawSpecialBuildings("man_made", "tower", 25, "man_made_tower", "building_nonliving_roof1");
	osm2pov_converter.drawSpecialBuildings("amenity", "tower", 25, "man_made_tower", "building_nonliving_roof1");
	osm2pov_converter.drawSpecialBuildings("man_made", "chimney", 50, "man_made_tower", NULL);
	osm2pov_converter.drawWays("barrier", "wall", 0.3, 3, "wall", true, false);

	osm2pov_converter.drawObjects("power_source", "wind", "windpower", 1.5, 1, 1);
	osm2pov_converter.drawObjects("amenity", "post_box", "postbox", 0.1, 1, 1);
	osm2pov_converter.drawObjects("natural", "tree", "tree", 0.2, 1, 6);
	osm2pov_converter.drawForests("natural", "wood", 0.00011, "forest", "tree", 1, 1, 6);
	osm2pov_converter.drawForests("landuse", "forest", 0.00011, "forest", "tree", 1, 1, 6);
	osm2pov_converter.drawForests("landuse", "wood", 0.00011, "forest", "tree", 1, 1, 6);

	osm2pov_converter.drawAreas("nature", "scrub", 0.00012, "greenplace");
	osm2pov_converter.drawForests("leisure", "park", 0.00012, "greenplace", "tree", 1, 1, 6);
	osm2pov_converter.drawForests("leisure", "garden", 0.00012, "greenplace", "tree", 1, 1, 6);
	osm2pov_converter.drawAreas("landuse", "village_green", 0.00012, "greenplace");
	osm2pov_converter.drawAreas("landuse", "allotments", 0.00012, "greenplace");
	osm2pov_converter.drawAreas("landuse", "meadow", 0.00012, "greenplace");
	osm2pov_converter.drawAreas("landuse", "greenfield", 0.00012, "greenplace");

	osm2pov_converter.drawAreas("landuse", "residential", 0.00007, "landuse_residential");
	osm2pov_converter.drawAreas("landuse", "industrial", 0.00008, "landuse_industrial");
	osm2pov_converter.drawAreas("landuse", "farmland", 0.00009, "landuse_farmland");
	osm2pov_converter.drawAreas("landuse", "farm", 0.00009, "landuse_farmland");
	osm2pov_converter.drawAreas("landuse", "farmyard", 0.00009, "landuse_farmland");

	osm2pov_converter.drawWays("waterway", "drain", 1, 0.00013, "river", true, true);
	osm2pov_converter.drawWays("waterway", "stream", 2, 0.00013, "river", true, true);
	osm2pov_converter.drawWays("waterway", "canal", 2.5, 0.00013, "river", true, true);
	osm2pov_converter.drawWays("waterway", "river", 5, 0.00013, "river", true, true);
	osm2pov_converter.drawAreas("waterway", "riverbank", 0.00013, "river");
	osm2pov_converter.drawAreas("natural", "water", 0.00013, "river");
	osm2pov_converter.drawAreas("landuse", "basin", 0.00013, "river");
	osm2pov_converter.drawAreas("landuse", "reservoir", 0.00013, "river");

/*	{		//info about not used tags. Now isn't used :-)
		multimap<size_t,string> disused_attributes;
		primitives.getDisusedAttributes(&disused_attributes);
		if (!disused_attributes.empty()) {
			cout << "Disused attributes:" << endl;
			for (multimap<size_t,string>::reverse_iterator it = disused_attributes.rbegin(); it != disused_attributes.rend(); it++) {
				cout << it->first << "x " << it->second << endl;
			}
		}
	}
*/
	cout << "Done." << endl;
}
